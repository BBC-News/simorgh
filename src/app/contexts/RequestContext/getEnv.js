const getEnv = origin => {
  if (origin.includes('local')) {
    return 'local';
  }

  if (origin.includes('test')) {
    return 'test';
  }

  if (origin.includes('stage')) {
    return 'stage';
  }

  return 'live';
};

export default getEnv;
