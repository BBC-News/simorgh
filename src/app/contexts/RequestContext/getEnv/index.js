const getEnv = origin => {
  if (origin.includes('local')) {
    return 'local';
  }

  if (origin.includes('test')) {
    return 'test';
  }

  return 'live';
};

export default getEnv;
