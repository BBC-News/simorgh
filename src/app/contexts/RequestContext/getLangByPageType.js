import deepGet from '../../lib/utilities/deepGet';

const getLangByPageType = (data, defaultLang, pageType) => {
  return pageType === 'article'
    ? deepGet(['pageData', 'metadata', 'passport', 'language'], data)
    : defaultLang;
};

export default getLangByPageType;
