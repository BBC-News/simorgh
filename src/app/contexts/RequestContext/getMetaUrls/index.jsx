const getAmpUrl = url => `${url}.amp`;

const getCanonicalUrl = (origin, service, pageType, id) => {
  switch (pageType) {
    case 'frontPage':
      return `${origin}/${service}`;
    default:
      return `${origin}/${service}/articles/${id}`;
  }
};

const getCanonicalUrlWithVariant = (
  origin,
  service,
  serviceVariant,
  pageType,
  id,
) => {
  switch (pageType) {
    case 'frontPage':
      return `${origin}/${service}/${serviceVariant}`;
    default:
      return `${origin}/${service}/articles/${id}/${serviceVariant}`;
  }
};

const getUkCanonicalUrl = url => {
  const tldRegex = /(.com|.co.uk)/g;
  return url.replace(tldRegex, '.co.uk');
};

const getNonUkCanonicalUrl = url => {
  const tldRegex = /(.com|.co.uk)/g;
  return url.replace(tldRegex, '.com');
};

const getMetaUrls = (origin, service, serviceVariant, pageType, id) => {
  let canonicalLink = getCanonicalUrl(origin, service, pageType, id);
  let canonicalUkLink = getUkCanonicalUrl(canonicalLink);
  let canonicalNonUkLink = getNonUkCanonicalUrl(canonicalLink);

  if (serviceVariant) {
    canonicalLink = getCanonicalUrlWithVariant(
      origin,
      service,
      serviceVariant,
      pageType,
      id,
    );
    canonicalUkLink = getUkCanonicalUrl(canonicalLink);
    canonicalNonUkLink = getNonUkCanonicalUrl(canonicalLink);

    return {
      canonicalLink,
      ampLink: getAmpUrl(canonicalLink),
      canonicalUkLink,
      ampUkLink: getAmpUrl(canonicalUkLink),
      canonicalNonUkLink,
      ampNonUkLink: getAmpUrl(canonicalNonUkLink),
    };
  }

  return {
    canonicalLink,
    ampLink: getAmpUrl(canonicalLink),
    canonicalUkLink,
    ampUkLink: getAmpUrl(canonicalUkLink),
    canonicalNonUkLink,
    ampNonUkLink: getAmpUrl(canonicalNonUkLink),
  };
};

export default getMetaUrls;
