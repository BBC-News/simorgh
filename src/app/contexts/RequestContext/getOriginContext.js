import get from '../../helpers/json/deepGet';
import onClient from '../../helpers/onClient';

const getArticleID = articleData => {
  const aresID = get(['metadata', 'id'], articleData);
  return aresID ? aresID.split(':').pop() : 'unknown';
};

const getOriginContext = (bbcOrigin, service, articleData) => {
  let origin = 'https://www.bbc.co.uk';
  let isUK = true;
  let referrer = null;
  let env = 'live';

  let href = `${origin}/${service}/${getArticleID(articleData)}`;

  if (bbcOrigin) {
    origin = bbcOrigin;
  }

  if (onClient()) {
    if (window.location.origin) {
      origin = window.location.origin; // eslint-disable-line prefer-destructuring
    }

    if (window.location.href) {
      href = window.location.href; // eslint-disable-line prefer-destructuring
    }

    if (document.referrer) {
      referrer = document.referrer; // eslint-disable-line prefer-destructuring
    }
  }

  if (origin.includes('.stage.')) {
    env = 'stage';
  }

  if (origin.includes('.test.')) {
    env = 'test';
  }

  if (origin.includes('localhost')) {
    env = 'local';
  }

  if (origin.includes('.com')) {
    isUK = false;
  }

  return {
    origin,
    isUK,
    env,
    href,
    referrer,
  };
};

export default getOriginContext;
