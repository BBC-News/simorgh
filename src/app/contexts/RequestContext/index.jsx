import React from 'react';
import { bool, node, oneOf, string } from 'prop-types';

export const RequestContext = React.createContext('default');

export const RequestContextProvider = ({
  children,
  env,
  id,
  isUK,
  lang,
  origin,
  pageType,
  platform,
  statsDestination,
  statsPageIdentifier,
}) => {
  const value = {
    env,
    id,
    isUK,
    lang,
    origin,
    pageType,
    platform,
    statsDestination,
    statsPageIdentifier,
  };

  return (
    <RequestContext.Provider value={value}>{children}</RequestContext.Provider>
  );
};

RequestContextProvider.propTypes = {
  children: node.isRequired,
  env: string,
  id: string,
  lang: string.isRequired,
  pageType: oneOf(['article', 'frontPage']).isRequired,
  platform: string.isRequired,
  isUK: bool.isRequired,
  origin: string.isRequired,
  statsDestination: string.isRequired,
  statsPageIdentifier: string.isRequired,
};

RequestContextProvider.defaultProps = {
  id: null,
  env: 'local',
};
