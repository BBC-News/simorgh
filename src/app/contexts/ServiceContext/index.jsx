import React from 'react';
import { node, string } from 'prop-types';
import services from '#lib/config/services/loadableConfig';
import variantPropType from '../../models/propTypes/variants';
import { getVariant } from '#lib/utilities/variantHandler';

export const ServiceContext = React.createContext({});

export const ServiceContextProvider = ({ children, service, variant }) => {
  const LoadableContextProvider = services[service];

  if (!LoadableContextProvider) {
    return null;
  }

  return (
    <LoadableContextProvider
      Context={ServiceContext}
      dataKey={getVariant({ service, variant })}
    >
      {children}
    </LoadableContextProvider>
  );
};

ServiceContextProvider.propTypes = {
  children: node.isRequired,
  service: string,
  variant: variantPropType,
};

ServiceContextProvider.defaultProps = {
  service: 'default',
  variant: 'default',
};
