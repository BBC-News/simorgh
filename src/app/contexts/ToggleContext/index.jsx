import React, { createContext, useEffect, useReducer } from 'react';
import { node, string } from 'prop-types';
import nodeLogger from '#lib/logger.node';
import { TOGGLE_FETCH_ERROR } from '#lib/logger.const';
import { toggleReducer, updateToggles } from './reducer';
import defaultToggles from '#lib/config/toggles';
import constructTogglesEndpoint from './utils/constructTogglesEndpoint';

const logger = nodeLogger(__filename);
const ToggleContext = createContext({});

const ToggleContextProvider = ({ children, service, origin }) => {
  const environment = process.env.SIMORGH_APP_ENV || 'local';
  const simorghToggles = defaultToggles[environment];
  const [toggleState, toggleDispatch] = useReducer(
    toggleReducer,
    simorghToggles,
  );

  // temp method to only enable remote feature toggling for test and for a list of services
  const { enableFetchingToggles } = simorghToggles;

  useEffect(() => {
    const shouldFetchAndUpdateToggles =
      enableFetchingToggles.enabled &&
      RegExp(enableFetchingToggles.value).test(service);

    if (shouldFetchAndUpdateToggles) {
      const fetchAndUpdateToggles = async () => {
        try {
          const response = await fetch(
            constructTogglesEndpoint(service, origin),
          );

          const jsonData = await response.json();

          toggleDispatch(updateToggles(jsonData));
        } catch (error) {
          logger.error(TOGGLE_FETCH_ERROR, { error });
        }
      };
      fetchAndUpdateToggles();
    }
  }, [
    service,
    origin,
    enableFetchingToggles.enabled,
    enableFetchingToggles.value,
    simorghToggles,
  ]);

  return (
    <ToggleContext.Provider value={{ toggleState, toggleDispatch }}>
      {children}
    </ToggleContext.Provider>
  );
};

const ToggleContextConsumer = ToggleContext.Consumer;

ToggleContextProvider.propTypes = {
  children: node.isRequired,
  origin: string.isRequired,
  service: string.isRequired,
};

export { ToggleContext, ToggleContextProvider, ToggleContextConsumer };
