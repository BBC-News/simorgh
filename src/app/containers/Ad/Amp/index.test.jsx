import React from 'react';
import { shouldMatchSnapshot } from '@bbc/psammead-test-helpers';
import { render } from '@testing-library/react';
import AmpAd, { AMP_ACCESS_FETCH } from './index';
import { ServiceContextProvider } from '#contexts/ServiceContext';
import { RequestContextProvider } from '#contexts/RequestContext';
import { ToggleContext } from '#contexts/ToggleContext';

const adJsonAttributes = slotType => ({
  targeting: {
    slot: slotType,
    asset_type: 'index',
    channel: 'afrique',
  },
});

const adWithContext = (slotType, toggleContext) => (
  <RequestContextProvider
    bbcOrigin="https://www.test.bbc.com"
    isAmp
    pageType="frontPage"
    service="afrique"
    pathname="/"
  >
    <ServiceContextProvider service="afrique">
      <ToggleContext.Provider value={toggleContext}>
        <AmpAd slotType={slotType} />
      </ToggleContext.Provider>
    </ServiceContextProvider>
  </RequestContextProvider>
);

describe('AMP Ads', () => {
  beforeAll(() => {
    process.env.SIMORGH_CONFIG_URL = 'https://mock-toggles-endpoint.bbc.co.uk';
  });

  afterAll(() => {
    delete process.env.SIMORGH_CONFIG_URL;
  });

  describe('Snapshots', () => {
    const toggleState = {
      ampAds: {
        enabled: true,
      },
    };

    const mockToggleDispatch = jest.fn();

    const toggleContextMock = {
      toggleState,
      toggleDispatch: mockToggleDispatch,
    };

    shouldMatchSnapshot(
      'should correctly render an AMP leaderboard ad',
      adWithContext('leaderboard', toggleContextMock),
    );

    shouldMatchSnapshot(
      'should correctly render an AMP mpu ad',
      adWithContext('mpu', toggleContextMock),
    );
  });

  describe('Assertions', () => {
    describe('When ampAds is toggled on', () => {
      const toggleState = {
        ampAds: {
          enabled: true,
        },
      };

      const mockToggleDispatch = jest.fn();

      const toggleContextMock = {
        toggleState,
        toggleDispatch: mockToggleDispatch,
      };

      it('should render two leaderboard ads', () => {
        const { container } = render(
          adWithContext('leaderboard', toggleContextMock),
        );
        const ampAd = container.querySelectorAll('amp-ad');

        expect(ampAd.length).toBe(2);
      });

      it('should display leaderboard ad with values for all of the needed attributes', () => {
        const { container } = render(
          adWithContext('leaderboard', toggleContextMock),
        );

        const ampAd = container.querySelectorAll('amp-ad');
        ampAd.forEach(ad => {
          expect(ad).toHaveAttribute('data-block-on-consent', 'default');
          expect(ad).toHaveAttribute('data-npa-on-unknown-consent', 'true');
          expect(ad).toHaveAttribute('media');
          expect(ad).toHaveAttribute('type');
          expect(ad).toHaveAttribute('width');
          expect(ad).toHaveAttribute('height');
          expect(ad).toHaveAttribute('data-multi-size');
          expect(ad).toHaveAttribute('data-slot');
          expect(ad).toHaveAttribute('data-amp-slot-index');
          expect(ad).toHaveAttribute('data-a4a-upgrade-type');
          expect(ad).toHaveAttribute(
            'json',
            JSON.stringify(adJsonAttributes('leaderboard')),
          );
        });
      });

      it('should render one mpu ad', () => {
        const { container } = render(adWithContext('mpu', toggleContextMock));
        const ampAd = container.querySelectorAll('amp-ad');

        expect(ampAd.length).toBe(1);
      });

      it('should display mpu ad with values for all of the needed attributes', () => {
        const { container } = render(adWithContext('mpu', toggleContextMock));

        const ampAd = container.querySelectorAll('amp-ad');
        ampAd.forEach(ad => {
          expect(ad).toHaveAttribute('data-block-on-consent', 'default');
          expect(ad).toHaveAttribute('data-npa-on-unknown-consent', 'true');
          expect(ad).toHaveAttribute('type');
          expect(ad).toHaveAttribute('width');
          expect(ad).toHaveAttribute('height');
          expect(ad).toHaveAttribute('data-multi-size');
          expect(ad).toHaveAttribute('data-slot');
          expect(ad).toHaveAttribute('data-amp-slot-index');
          expect(ad).toHaveAttribute('data-a4a-upgrade-type');
          expect(ad).toHaveAttribute(
            'json',
            JSON.stringify(adJsonAttributes('mpu')),
          );
        });
      });

      it('should render an `advertisement` label', () => {
        const { container } = render(
          adWithContext('leaderboard', toggleContextMock),
        );
        const links = container.querySelectorAll('a');
        const advertisementLabel = links && links[0];
        expect(advertisementLabel.textContent).toEqual('PublicitÃ©s');
        expect(advertisementLabel).toHaveAttribute('tabIndex', '-1');
      });
    });

    describe('When ampAds is toggled off', () => {
      const toggleState = {
        ampAds: {
          enabled: false,
        },
      };

      const mockToggleDispatch = jest.fn();

      const toggleContextMock = {
        toggleState,
        toggleDispatch: mockToggleDispatch,
      };

      it('should not render mpu ads on AMP when ampAds toggle is set to false', () => {
        const { container } = render(adWithContext('mpu', toggleContextMock));

        expect(container.firstChild).toBeNull();
      });

      it('should not render leaderboard ads on AMP when ampAds toggle is set to false', () => {
        const { container } = render(
          adWithContext('leaderboard', toggleContextMock),
        );

        expect(container.firstChild).toBeNull();
      });
    });

    describe('AMP_ACCESS_FETCH', () => {
      it('should retrieve data from the correct endpoint', () => {
        const ampAccessFetch = jest.fn().mockImplementation(AMP_ACCESS_FETCH);
        const ampAccessData = ampAccessFetch('afrique');
        const expectedReturn =
          'https://mock-toggles-endpoint.bbc.co.uk?application=simorgh&service=afrique';

        expect(ampAccessFetch).toHaveReturned();
        expect(ampAccessFetch).toHaveBeenCalledWith('afrique');
        expect(ampAccessData.type).toEqual('script');
        expect(JSON.parse(ampAccessData.props.children).authorization).toEqual(
          expectedReturn,
        );
      });
    });
  });
});
