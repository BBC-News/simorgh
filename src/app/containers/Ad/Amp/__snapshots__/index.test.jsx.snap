// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AMP Navigation Snapshots should correctly render an AMP leaderboard ad 1`] = `
.c0 {
  margin: 0 auto;
  width: 100%;
  text-align: center;
}

@media (min-width:37.5rem) {
  .c0 {
    margin-top: 1.5rem;
  }
}

@media (min-width:63rem) {
  .c0 {
    margin-top: 2rem;
    max-width: 63rem;
  }
}

<html>
  <head>
    <script
      async="true"
      custom-element="amp-ad"
      src="https://cdn.ampproject.org/v0/amp-ad-0.1.js"
    />
    <script
      async="true"
      custom-element="amp-access"
      src="https://cdn.ampproject.org/v0/amp-access-0.1.js"
    />
    <script
      id="amp-access"
      type="application/json"
    >
      {"authorization":"https://mock-toggles-endpoint.bbc.co.uk/toggles?application=simorgh&service=pidgin&geoiplookup=true","noPingback":true}
    </script>
  </head>
  <body>
    <div>
      <div
        amp-access="toggles.ads.enabled AND geoIp.advertiseCombined"
        amp-access-hide="true"
      >
        <div
          class="c0"
        >
          <amp-ad
            data-a4a-upgrade-type="amp-ad-network-doubleclick-impl"
            data-amp-slot-index="0"
            data-multi-size="320x50,300x50"
            data-slot="/4817/bbccom.test.site.amp.news"
            height="50"
            json="{\\"targeting\\":{\\"slot\\":\\"leaderboard\\",\\"asset_type\\":\\"index\\",\\"channel\\":\\"pidgin\\"}}"
            media="(max-width: 599px)"
            type="doubleclick"
            width="320"
          >
            <amp-img
              height="50"
              layout="responsive"
              placeholder="true"
              src="https://via.placeholder.com/320x50"
              width="320"
            />
          </amp-ad>
          <amp-ad
            data-a4a-upgrade-type="amp-ad-network-doubleclick-impl"
            data-amp-slot-index="0"
            data-multi-size="970x250,728x90"
            data-slot="/4817/bbccom.test.site.amp.news"
            height="250"
            json="{\\"targeting\\":{\\"slot\\":\\"leaderboard\\",\\"asset_type\\":\\"index\\",\\"channel\\":\\"pidgin\\"}}"
            media="(min-width: 600px)"
            type="doubleclick"
            width="970"
          >
            <amp-img
              height="250"
              layout="responsive"
              placeholder="true"
              src="https://via.placeholder.com/970x250"
              width="970"
            />
          </amp-ad>
        </div>
      </div>
    </div>
  </body>
</html>
`;
